cassiopeia/solver/SpectralNeighborJoiningSolver.py

```diff
@@ -0,0 +1,318 @@
+        a_comp_flat = list(itertools.chain.from_iterable(a_comp))
+
+        # reconstruct RA matrix
+        RA_matrix = self._similarity_map.values[np.ix_(a_subset, a_comp_flat)]
+
+        # get second largest SVD if available, first if not.
+        s = svd(RA_matrix, compute_uv=False, check_finite=False)
+        svd2_val = s[:2][-1]
+
+        return svd2_val
+
+    def update_dissimilarity_map(
+        self,
+        similarity_map: pd.DataFrame,
+        cherry: Tuple[str, str],
+        new_node: str,
+    ) -> pd.DataFrame:
+        """Updates the lambda matrix using the pair of subsets from find_cherry.
+
+        Args:
+            similarity_map: lambda matrix to update
+            cherry1: One of the children to join.
+            cherry2: One of the children to join.
+            new_node: New node name to add to the dissimilarity map
+
+        Returns:
+            An updated lambda matrix.
+        """
+        # get cherry nodes in index of lambda matrix
+        i, j = (
+            np.where(similarity_map.index == cherry[0])[0][0],
+            np.where(similarity_map.index == cherry[1])[0][0],
+        )
+
+        # modify names
+        node_names = similarity_map.index.values
+        node_names = np.append(node_names, new_node)
+        node_names = np.delete(node_names, [i, j])
+
+        # modify indices
+        self.lambda_indices.append(
+            [*self.lambda_indices[i], *self.lambda_indices[j]]
+        )
+        self.lambda_indices.pop(max(i, j))
+        self.lambda_indices.pop(min(i, j))
+
+        # new lambda indices
+        N = len(self.lambda_indices)
+
+        if N <= 2:
+            return pd.DataFrame(
+                np.zeros([N, N]), index=node_names, columns=node_names
+            )
+
+        # get the old lambda matrix
+        lambda_matrix_arr = similarity_map.drop(
+            index=[cherry[0], cherry[1]], columns=[cherry[0], cherry[1]]
+        ).values
+
+        # add new col + row to lambda matrix
+        new_row = np.array([0.0] * (N - 1))
+        lambda_matrix_arr = np.vstack(
+            (lambda_matrix_arr, np.atleast_2d(new_row))
+        )
+        new_col = np.array([0.0] * N)
+        lambda_matrix_arr = np.array(
+            np.hstack((lambda_matrix_arr, np.atleast_2d(new_col).T))
+        )
+
+        # compute new SVDs
+        i_idx = N - 1
+        for j_idx in range(i_idx):
+            svd2_val = self._compute_svd2(
+                pair=(i_idx, j_idx), lambda_indices=self.lambda_indices
+            )
+
+            lambda_matrix_arr[i_idx, j_idx] = lambda_matrix_arr[
+                j_idx, i_idx
+            ] = svd2_val
+
+        np.fill_diagonal(lambda_matrix_arr, np.inf)
+
+        # regenerate lambda matrix
+        lambda_matrix_df = pd.DataFrame(
+            lambda_matrix_arr, index=node_names, columns=node_names
+        )
+
+        return lambda_matrix_df
+
+    def setup_root_finder(self, cassiopeia_tree: CassiopeiaTree) -> None:
```

Comment:
I wonder if we should make this the default function in `DistanceSolver`, which can be overidden by solvers like `UPGMASolver`. It just tends to be poor practice to duplicate code like this across subclasses. What do you think?

Reply:
I might be biased since I worked on SNJ and not UPGMASolver, but I do think that having it in DistanceSolver to be overridden is a good idea, like the new get_dissimilarity_map method.

Reply:
Thanks for the suggestion, @ivalexander13 -- @richardyz98  and @sprillo , what do you two think about this?

Reply:
I personally believe that the `DistanceSolver` class should not have a default function, as I believe that it is not the case that this behavior of adding an implicit root is inherent to DistanceSolvers, with `UPGMASolver` being a special case. Let me know your thoughts all!

Reply:
That's fine, let's keep it as is for now. Thanks Richard!